src/simulators/statevector/qubitvector_thrust.hpp

```diff
@@ -79,6 +89,722 @@ using indexes_t = std::unique_ptr<uint_t[]>;
 template <size_t N> using areg_t = std::array<uint_t, N>;
 template <typename T> using cvector_t = std::vector<std::complex<T>>;
 
+//==================================
+// parameters for gate kernels
+//==================================
+template <typename data_t>
+struct GateParams {
+  thrust::complex<data_t>* buf_;
+  uint_t* offsets_;
+  thrust::complex<double>* matrix_;
+  uint_t* params_;
+  uint_t gid_;
+  uint_t lmask_;
+};
+
+//========================================
+//  base class of gate functions
+//========================================
+class GateFuncBase
+{
+public:
+  GateFuncBase(void)
+  {
+  }
+
+  virtual bool IsDiagonal(void)
+  {
+    return false;
+  }
+  virtual int NumControlBits(void)
+  {
+    return 0;
+  }
+  virtual int ControlMask(void)
+  {
+    return 1;
+  }
+  virtual bool Reduction(void)
+  {
+    return false;
+  }
+
+  virtual const char* Name(void)
+  {
+    return "base function";
+  }
+};
+
+
+//=============================================================
+//    virtual buffer class
+//=============================================================
+template <typename data_t>
+class QubitVectorBuffer
+{
+protected:
+  uint_t m_size;
+public:
+  QubitVectorBuffer(uint_t size = 0)
+  {
+    m_size = size;
+  }
+  uint_t Size(void)
+  {
+    return m_size;
+  }
+
+  virtual data_t* BufferPtr(void)
+  {
+    return NULL;
+  }
+  virtual void Set(uint_t i,const data_t& t) = 0;
+  virtual data_t Get(uint_t i) const = 0;
+
+  virtual void Resize(uint_t size) = 0;
+
+  virtual void Copy(const std::vector<data_t>& v) = 0;
+
+  virtual void Copy(uint_t pos,QubitVectorBuffer<data_t>* pSrc,uint_t srcPos,uint_t size,int isDevice = 0) = 0;
+
+  virtual void CopyIn(uint_t pos,const data_t* pSrc,uint_t size) = 0;
+  virtual void CopyOut(uint_t pos,data_t* pDest,uint_t size) = 0;
+
+};
+
+#ifdef AER_THRUST_CUDA
+#define AERDeviceVector thrust::device_vector
```

Comment:
This would be clearer at the top of the file containing buffer classes using `using`

```cpp
#ifdef AER_THRUST_CUDA
using device_vector = thrust::device_vector
#else 
using device_vector = thrust::host_vector
#endif
```